### PR TITLE
[runtime] Handle circular references in SRE types

### PR DESCRIPTION
**Pull request by alexanderkyte**
We previously recursed through the entire directed cyclic graph
of the type passed to SRE. When given a (legal) circular type reference,
we encountered stack overflows. This was particularly easy to trigger
with fsharpi and with FAKE.

By creating everything about the type before searching for the parent,
and by defering all type hierarchy establishment until the type must be
surfaced to manage code, we tease out the data dependencies caught up in
the cycle.

Note the bounded lifetime of the unmanaged object hidden in the SRE
module, which acts as a global reference for all types reachable from
the TypeBuilder passed to the runtime through the icall.

### CODE REVIEW

File: mono/metadata/object-internals.h [hunk 1]
```diff
@@ -1182,11 +1182,18 @@ typedef struct {
 	MonoArray *global_fields;
 	gboolean is_main;
 	MonoArray *resources;
+	GHashTable *unparented_classes;
```
Comment (lambdageek) [inline mono/metadata/object-internals.h]: You also need to change `mcs/class/corlib/System.Reflection.Emit/ModuleBuilder.cs`

File: mono/metadata/object-internals.h [hunk 2]
```diff
@@ -1212,6 +1219,7 @@ typedef struct {
 	MonoArray *generic_params;
 	MonoArray *permissions;
 	MonoReflectionType *created;
+	gint32 state;
```
Comment (lambdageek) [inline mono/metadata/object-internals.h]: Also update `mcs/class/corlib/System.Reflection.Emit/TypeBuilder.cs`

File: mono/metadata/sre.c [hunk 3]
```diff
@@ -2336,52 +2337,65 @@ mono_reflection_get_custom_attrs_blob_checked (MonoReflectionAssembly *assembly,
...
+	mono_loader_lock ();
+
+	gint32 entering_state = MONO_HANDLE_GETVAL (MONO_HANDLE_CAST (MonoReflectionTypeBuilder, ref_tb), state);
+	if (entering_state != MonoTypeBuilderNew) {
+		g_assert (MONO_HANDLE_GETVAL (MONO_HANDLE_CAST (MonoReflectionType, ref_tb), type));
 		goto leave;
 	}
 
+	MONO_HANDLE_SETVAL (ref_tb, state, MonoTypeBuilderState, MonoTypeBuilderEntered);
+	MonoReflectionModuleBuilder *module = MONO_HANDLE_GETVAL (ref_tb, module);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: `MonoReflectionModuleBuilderHandle module = MONO_HANDLE_NEW_GET (MonoReflectionModuleBuilder, ref_tb, module);`

File: mono/metadata/sre.c [hunk 4]
```diff
@@ -2457,9 +2461,45 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
+
+	// Do here so that the search inside of the parent can see the above type that's been set.
+	MonoReflectionTypeHandle ref_parent = MONO_HANDLE_NEW_GET (MonoReflectionType, ref_tb, parent);
+	MonoType *parent_type = NULL;
+	if (!MONO_HANDLE_IS_NULL (ref_parent)) {
+		MonoClass *parent_klass = mono_handle_class (ref_parent);
+		gboolean recursive_init = TRUE;
+
+		if (is_sre_type_builder (parent_klass)) {
+			MonoTypeBuilderState parent_state = MONO_HANDLE_GETVAL (MONO_HANDLE_CAST (MonoReflectionTypeBuilder, ref_tb), state);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: Shouldn't `ref_tb` here be `ref_parent` ?

File: mono/metadata/sre.c [hunk 5]
```diff
@@ -2457,9 +2461,45 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
+
+			if (parent_state != MonoTypeBuilderNew) {
+				// Initialize types reachable from parent recursively
+				// We'll fix the type hierarchy later
+				recursive_init = FALSE;
+			}
+		}
+
+		if (recursive_init)
+			mono_reflection_type_handle_mono_type (ref_parent, error);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: What does this do?

File: mono/metadata/sre.c [hunk 6]
```diff
@@ -2481,24 +2521,23 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
  *
  * Creates the generic class after all generic parameters have been added.
  * On success returns TRUE, on failure returns FALSE and sets @error.
+ *
+ * This assumes that reflection_setup_internal_class has already set up
+ * ref_tb
  */
 static gboolean
-reflection_create_generic_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
+reflection_inflate_generic_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
```
Comment (lambdageek) [inline mono/metadata/sre.c]: I don't think "inflate" is a great name for this.  I think the runtime usually uses "inflate" to mean "instantiate", but this function isn't creating a generic instance, it's creating the generic context for a generic type definition.

File: mono/metadata/sre.c [hunk 7]
```diff
@@ -2481,24 +2521,23 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
 {
 	HANDLE_FUNCTION_ENTER ();
 
 	mono_error_init (error);
 
-	reflection_setup_internal_class (ref_tb, error);
-	if (!is_ok (error))
-		goto leave;
-
```
Comment (lambdageek) [inline mono/metadata/sre.c]: Can we assert anything here to make sure it's not called too early?

File: mono/metadata/sre.c [hunk 8]
```diff
@@ -2432,16 +2446,6 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
-	if (parent != NULL) {
-		mono_class_setup_parent (klass, parent);
-	} else if (strcmp (klass->name, "Object") == 0 && strcmp (klass->name_space, "System") == 0) {
-		const char *old_n = klass->name;
-		/* trick to get relative numbering right when compiling corlib */
-		klass->name = "BuildingObject";
-		mono_class_setup_parent (klass, mono_defaults.object_class);
-		klass->name = old_n;
-	}
-
```
Comment (lambdageek) [inline mono/metadata/sre.c]: After you remove this, the call to `mono_class_setup_supertypes (klass)` on 2459 will do something broken.

File: mono/metadata/sre.c (hunk 5 above)
Comment (lambdageek) [inline mono/metadata/sre.c]: 1. Oh, I get it, this is done for the side-effect of creating the parent.  Could you add a comment to that effect?

2.  Not checking `error`.

File: mono/metadata/object-internals.h (hunk 2 above)
Comment (alexanderkyte) [inline mono/metadata/object-internals.h]: Oops, those were in the original PR. Let me add them.

File: mono/metadata/sre.c [hunk 9]
```diff
@@ -3526,12 +3565,49 @@ remove_instantiations_of_and_ensure_contents (gpointer key,
...
+		module->unparented_classes = unparented_classes;
+
+		gboolean ret_val = reflection_setup_internal_class_internal (ref_tb, error);
+		mono_error_assert_ok (error);
+
+		// Fix the relationship between the created classes and their parents
+		reflection_setup_class_hierarchy (ref_tb, unparented_classes, error);
+		mono_error_assert_ok (error);
+
+		module->unparented_classes = NULL;
```
Comment (kumpera) [inline mono/metadata/sre.c]: Who frees the hashtable?

**Comment by kumpera** [comment]
We need to reduce a test case from F#/Fake for this, otherwise it makes reviewing quite awkward.

**Comment by kumpera** [comment]
This PR is aborting the test suite with: `* Assertion at sre.c:2386, condition `MONO_HANDLE_GETVAL (MONO_HANDLE_CAST (MonoReflectionType, ref_tb), type)' not met`

**Comment by alexanderkyte** [comment]
Should have marked as WIP after aleksey made apparent that something had happened with the PR. Commits were lost because the linker external module was corrupted during rebase ("not porcelain?"), and then lost in the squash. Fixed, but testing locally.

**Comment by alexanderkyte** [comment]
I'll get an isolated reproduction

File: mono/metadata/object-internals.h (hunk 1 above)
Comment (alexanderkyte) [inline mono/metadata/object-internals.h]: I've fixed this, commit was missing due to internal git issue.

File: mono/metadata/object-internals.h (hunk 2 above)
Comment (alexanderkyte) [inline mono/metadata/object-internals.h]: I've fixed this, commit was missing due to internal git issue.

File: mono/metadata/sre.c (hunk 8 above)
Comment (alexanderkyte) [inline mono/metadata/sre.c]: I've moved this until after the class hierarchy init

File: mono/metadata/sre.c [hunk 10]
```diff
@@ -3526,12 +3571,50 @@ remove_instantiations_of_and_ensure_contents (gpointer key,
...
+ * Creates a MonoClass that represents the TypeBuilder.
+ * This is a trick that lets us simplify a lot of reflection code
+ * (and will allow us to support Build and Run assemblies easier).
+ *
+ * Returns TRUE on success. On failure, returns FALSE and sets @error.
+ */
+static gboolean
+reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
+{
+	MonoReflectionModuleBuilder *module = MONO_HANDLE_GETVAL (ref_tb, module);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: should be a handle for the module: `MonoReflectionModuleBuilderHandle module = MONO_HANDLE_NEW_GET (...)`

File: mono/metadata/sre.c [hunk 11]
```diff
@@ -2452,14 +2457,52 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
+		}
+
+		if (recursive_init) {
+			// If we haven't encountered a cycle, force the creation of ref_parent's type
+			mono_reflection_type_handle_mono_type (ref_parent, error);
+			if (!is_ok (error))
+				goto leave;
+		}
+
+		parent_type = MONO_HANDLE_GETVAL (MONO_HANDLE_CAST (MonoReflectionType, ref_parent), type);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: don't need a cast here, ref_parent is a MonoReflectionTypeHandle already

File: mono/metadata/sre.c [hunk 12]
```diff
@@ -2481,23 +2524,25 @@ reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoErr
...
+reflection_init_generic_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
 {
 	HANDLE_FUNCTION_ENTER ();
 
 	mono_error_init (error);
 
-	reflection_setup_internal_class (ref_tb, error);
-	if (!is_ok (error))
-		goto leave;
+	MonoTypeBuilderState ref_state = MONO_HANDLE_GETVAL (MONO_HANDLE_CAST (MonoReflectionTypeBuilder, ref_tb), state);
```
Comment (lambdageek) [inline mono/metadata/sre.c]: Also no need for a cast here

File: mono/metadata/sre.c (hunk 10 above)
Comment (alexanderkyte) [inline mono/metadata/sre.c]: Ah, thanks

**Comment by alexanderkyte** [comment]
build

File: mono/metadata/sre.c [hunk 13]
```diff
@@ -2336,52 +2337,66 @@ mono_reflection_get_custom_attrs_blob_checked (MonoReflectionAssembly *assembly,
...
- *
- * Creates a MonoClass that represents the TypeBuilder.
- * This is a trick that lets us simplify a lot of reflection code
- * (and will allow us to support Build and Run assemblies easier).
- *
- * Returns TRUE on success. On failure, returns FALSE and sets @error.
- */
 static gboolean
-reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
+reflection_setup_class_hierarchy (MonoReflectionTypeBuilderHandle ref_tb, GHashTable *unparented, MonoError *error)
```
Comment (lambdageek) [inline mono/metadata/sre.c]: This doesn't use ref_tb anymore, AFAICT. And it doesn't allocate new handles to do its work, so it doesn't need handle function enter/return, either.

**Pull request merged**